instrumentation-genai/opentelemetry-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/dict_util.py

```diff
@@ -0,0 +1,222 @@
+# limitations under the License.
+
+
+import json
+from typing import Any, Callable, Dict, Optional, Sequence, Set, Union
+
+Primitive = Union[bool, str, int, float]
+BoolList = list[bool]
+StringList = list[str]
+IntList = list[int]
+FloatList = list[float]
+HomogenousPrimitiveList = Union[BoolList, StringList, IntList, FloatList]
+FlattenedValue = Union[Primitive, HomogenousPrimitiveList]
+FlattenedDict = Dict[str, FlattenedValue]
+
+
+def _concat_key(prefix: Optional[str], suffix: str):
+    if not prefix:
+        return suffix
+    return f"{prefix}.{suffix}"
+
+
+def _is_primitive(v):
+    for t in [str, bool, int, float]:
+        if isinstance(v, t):
+            return True
+    return False
+
+
+def _is_homogenous_primitive_list(v):
+    if not isinstance(v, list):
+        return False
+    if len(v) == 0:
+        return True
+    if not _is_primitive(v[0]):
+        return False
+    first_entry_value_type = type(v[0])
+    for entry in v[1:]:
+        if not isinstance(entry, first_entry_value_type):
+            return False
+    return True
+
+
+def _get_flatten_func(
+    flatten_functions: Dict[str, Callable], key_names: set[str]
+):
+    for key in key_names:
+        flatten_func = flatten_functions.get(key)
+        if flatten_func is not None:
+            return flatten_func
+    return None
+
+
+def _flatten_compound_value(
+    key: str,
+    value: Any,
+    exclude_keys: Set[str],
+    rename_keys: Dict[str, str],
+    flatten_functions: Dict[str, Callable],
+    key_names: Set[str],
+    _from_json=False,
+) -> FlattenedDict:
+    flatten_func = _get_flatten_func(flatten_functions, key_names)
+    if flatten_func is not None:
+        func_output = flatten_func(
+            key,
+            value,
+            exclude_keys=exclude_keys,
+            rename_keys=rename_keys,
+            flatten_functions=flatten_functions,
+        )
+        if func_output is not None:
+            return {key: func_output}
+    if isinstance(value, dict):
+        return _flatten_dict(
+            value,
+            key_prefix=key,
+            exclude_keys=exclude_keys,
+            rename_keys=rename_keys,
+            flatten_functions=flatten_functions,
+        )
+    if isinstance(value, list):
+        if _is_homogenous_primitive_list(value):
+            return {key: value}
+        return _flatten_list(
+            value,
+            key_prefix=key,
+            exclude_keys=exclude_keys,
+            rename_keys=rename_keys,
+            flatten_functions=flatten_functions,
+        )
+    if hasattr(value, "model_dump"):
+        return _flatten_dict(
+            value.model_dump(),
+            key_prefix=key,
+            exclude_keys=exclude_keys,
+            rename_keys=rename_keys,
+            flatten_functions=flatten_functions,
+        )
+    if _from_json:
+        raise ValueError(
+            f"Cannot flatten value with key {key}; value: {value}"
+        )
```

Comment:
It looks like this causes the whole call to `flatten_dict()` to fail and no attributes would get added to the span

Reply:
Yes. The `flatten_dict` either succeeds or it does not. I think it is better to fail fully, loudly with a clear error than to partially succeed, making it difficult to debug why certain information is absent.

Reply:
Thanks for that pointer. I've updated this logic to fail more gracefully and to log an error instead of raising.